R/forecastingdata_org.R

```diff
@@ -0,0 +1,380 @@
+          forecast_horizon <- as.numeric(line[2])
+        else if(line[1] == "@missing")
+          contain_missing_values <- as.logical(line[2])
+        else if(line[1] == "@equallength")
+          contain_equal_length <- as.logical(line[2])
+      }
+    }
+    line <- readLines(file, n = 1)
+  }
+  
+  if(length(line) == 0)
+    stop("Missing data section.")
+  if(is.null(col_names))
+    stop("Missing attribute section.")
+  
+  line <- readLines(file, n = 1)
+  
+  if(length(line) == 0)
+    stop("Missing series information under data section.")
+  
+  for(col in col_names)
+    assign(col, NULL)
+  
+  values <- NULL
+  row_count <- 0
+  
+  # Get data
+  while(length(line) != 0){
+    full_info <- strsplit(line, ":")[[1]]
+    
+    if(length(full_info) != length(col_names)+1)
+      stop("Missing attributes/values in series.")
+    
+    series <- strsplit(tail(full_info, 1), ",")[[1]]
+    series[which(series == "?")] <- NA
+    series <- as.numeric(series)
+    
+    if(sum(is.na(series)) == length(series))
+      stop("All series values are missing. A given series should contains a set of comma separated numeric values. At least one numeric value should be there in a series.")
+    
+    values <- c(values, series)
+    row_count <- row_count + length(series)
+    
+    attributes <- head(full_info, length(full_info)-1)
+    
+    for(col in seq_along(col_names)){
+      
+      att <- eval(parse(text=col_names[col]))
+      
+      # This format supports 3 attribute types: string, numeric and date
+      if(col_types[col] == "date"){
+        if(is.null(frequency))
+          stop("Frequency is missing.")
+        else{
+          if(frequency %in% LOW_FREQUENCIES)
+            start_time <- as.POSIXct(attributes[col], format = "%Y-%m-%d %H-%M-%S")
+          else if(frequency %in% HIGH_FREQUENCIES)
+            start_time <- as.Date(attributes[col], format = "%Y-%m-%d %H-%M-%S")
+          else
+            stop("Invalid frequency.")
+          
+          if(is.na(start_time))
+            stop("Incorrect timestamp format. Specify your timestamps as YYYY-mm-dd HH-MM-SS")
+        }
+        
+        att <- append(att, seq(start_time, length = length(series), by = FREQ_MAP[[frequency]]))
+      }else{
+        if(col_types[col] == "numeric")
+          attributes[col] <- as.numeric(attributes[col])
+        else if(col_types[col] == "string")
+          attributes[col] <- as.character(attributes[col])
+        else
+          stop("Invalid attribute type.")
+        
+        if(is.na(attributes[col]))
+          stop("Invalid attribute values.")
+        
+        att <- append(att, rep(attributes[col], length(series)))
+      }
+      assign(col_names[col], att)
+    }
+    
+    line <- readLines(file, n = 1)
+  }
+  
+  data <- as.data.frame(matrix(nrow = row_count, ncol = length(col_names) + 1))
+  colnames(data) <- c(col_names, value_column_name)
+  
+  for(col in col_names)
+    data[[col]] <- eval(parse(text = col))
```

Comment:
This code is also dangerous, presumably needed to access the objects assigned in the environment above.
Instead, it is simpler to store the column values in a list for later use/access.

This is also problematic if the column name is non-syntactic, for instance if `col = "my-turnover"` - or worse, `col = "evil_function()"`.

Reply:
Fixed as suggested. Added attributes into a list.